File: Python/utils/crud_operations.py
import json
from pymongo import MongoClient
from pymongo.collection import Collection
from pymongo.results import UpdateResult, DeleteResult
from bson.raw_bson import RawBSONDocument
from typing import Optional, List, Union


class CrudOperations:
    @staticmethod
    def connect_to_client(uri: str) -> Optional[MongoClient]:
        try:
            return MongoClient(uri)
        except Exception as e:
            print(f"Connection Error: {e}")
            return None

    @staticmethod
    def close_client(client: MongoClient) -> None:
        try:
            client.close()
        except Exception as e:
            print(f"Close Client Error: {e}")

    @staticmethod
    def drop_database(client: MongoClient, db_name: str) -> None:
        try:
            client.drop_database(db_name)
        except Exception as e:
            print(f"Error dropping database '{db_name}': {e}")

    @staticmethod
    def get_collection(client: MongoClient, db_name: str, collection_name: str) -> Optional[Collection]:
        try:
            return client[db_name][collection_name]
        except Exception as e:
            print(f"Get Collection Error: {e}")
            return None

    @staticmethod
    def insert_document(collection: Collection, document: RawBSONDocument) -> None:
        try:
            collection.insert_one(document)
        except Exception as e:
            print(f"Insert document Error: {e}")

    @staticmethod
    def execute_query(collection: Collection, filter: Optional[dict] = None, sort: Optional[list] = None, projection: Optional[dict] = None, limit: Optional[int] = None) -> Optional[List[RawBSONDocument]]:
        try:
            query = collection.find(filter, projection)
            if sort is not None:
                query.sort(sort)
            if limit is not None and limit > 0:
                query.limit(limit)
            return list(query)
        except Exception as e:
            print(f"Query Execution Error: {e}")
            return None

    @staticmethod
    def execute_aggregate_query(collection: Collection, pipeline: List[dict]) -> Optional[List[RawBSONDocument]]:
        try:
            return list(collection.aggregate(pipeline))
        except Exception as e:
            print(f"Aggregate Query Execution Error: {e}")
            return None

    @staticmethod
    def update_document(collection: Collection, filter: dict, update: dict, options: Optional[dict] = None) -> Optional[UpdateResult]:
        try:
            return collection.update_one(filter, update, **options) if options else collection.update_one(filter, update)
        except Exception as e:
            print(f"Update document Error: {e}")
            return None

    @staticmethod
    def update_multiple_document(collection: Collection, filter: dict, update: dict, options: Optional[dict] = None) -> Optional[UpdateResult]:
        try:
            return collection.update_many(filter, update, **options) if options else collection.update_many(filter, update)
        except Exception as e:
            print(f"Update Multiple document Error: {e}")
            return None

    @staticmethod
    def delete_document(collection: Collection, filter: dict, options: Optional[dict] = None) -> Optional[DeleteResult]:
        try:
            return collection.delete_one(filter, **options) if options else collection.delete_one(filter)
        except Exception as e:
            print(f"Delete document Error: {e}")
            return None

    @staticmethod
    def delete_multiple_documents(collection: Collection, filter: dict, options: Optional[dict] = None) -> Optional[DeleteResult]:
        try:
            return collection.delete_many(filter, **options) if options else collection.delete_many(filter)
        except Exception as e:
            print(f"Delete Multiple documents Error: {e}")
            return None

    @staticmethod
    def write_to_output_file(query_str: str, content: str, file_path: str) -> None:
        with open(file_path, 'a') as f:
            f.write(f"{query_str}\n")
            f.write(f"{content}\n")
            f.write("\n\n\n")

    @staticmethod
    def print_query_result_to_file(query_str: str, result: Union[int, List[dict], UpdateResult, DeleteResult], file_path: str) -> None:
        if isinstance(result, int):
            # For count of documents
            content = f"Count: {result}"
        elif isinstance(result, list):
            # For a list of documents
            content = "[\n"
            for doc in result:
                content += "\t{\n"
                for key, value in doc.items():
                    content += f"\t\t{key}: {value}"
                    if doc[key] is not result[-1][key]:
                        content += ','
                    content += '\n'
                content += "\t},\n" if doc is not result[-1] else "\t}\n"
            content += "]"
        elif isinstance(result, UpdateResult):
            # For UpdateResult            
            content = "{\n"
            content += f"\t\"acknowledged\": {result.acknowledged},\n"
            content += f"\t\"insertedId\": {str(result.upserted_id) if result.upserted_id else 'null'},\n"
            content += f"\t\"matchedCount\": {result.matched_count},\n"
            content += f"\t\"modifiedCount\": {result.modified_count},\n"
            content += f"\t\"upsertedCount\": 0\n"
            content += "}"

        elif isinstance(result, DeleteResult):
            # For DeleteResult
            content = f"AcknowledgedDeleteResult"
            content += "{deletedCount="
            content += f"{result.deleted_count}"
            content += "}"
        else:
            content = "Unsupported result type"

        CrudOperations.write_to_output_file(query_str, content, file_path)

File: Go/utils/crud_operations.go
package utils

import (
	"bufio"
	"context"
	"fmt"
	"log"
	"os"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

func ConnectToClient(uri string) (*mongo.Client, error) {
	client, err := mongo.Connect(context.TODO(), options.Client().ApplyURI(uri))
	if err != nil {
		return nil, fmt.Errorf("connection error: %v", err)
	}
	return client, nil
}

func CloseClient(client *mongo.Client) {
	if err := client.Disconnect(context.TODO()); err != nil {
		log.Printf("Error disconnecting from MongoDB: %v", err)
	}
}

func DropDatabase(client *mongo.Client, dbName string) {
	db := client.Database(dbName)
	if err := db.Drop(context.TODO()); err != nil {
		log.Printf("Drop database error: %v", err)
	}
}

func GetCollection(client *mongo.Client, dbName string, collectionName string) *mongo.Collection {
	return client.Database(dbName).Collection(collectionName)
}

func InsertDocument(collection *mongo.Collection, document interface{}) {
	_, err := collection.InsertOne(context.TODO(), document)
	if err != nil {
		log.Printf("Insert document error: %v", err)
	}
}

func writeToOutputFile(queryStr, content, filePath string) {
	file, err := os.OpenFile(filePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	writer := bufio.NewWriter(file)
	_, _ = writer.WriteString(fmt.Sprintf("%s\n%s\n\n\n\n", queryStr, content))
	writer.Flush()
}

func PrintQueryResultToFile(queryStr string, result interface{}, filePath string) {
	var content string

	switch r := result.(type) {
	case int64:
		content = fmt.Sprintf("Count: %d", r)
	case *mongo.Cursor:
		content = "[\n"
		for r.Next(context.TODO()) {
			content += "\t{\n"
			var result bson.M
			err := r.Decode(&result)
			if err != nil {
				log.Fatal(err)
			}
			for k, v := range result {
				content += fmt.Sprintf("\t\t\"%s\": \"%v\",\n", k, v)
			}
			content += "\t},\n"
		}
		content += "]"
	case *mongo.UpdateResult:
		content = fmt.Sprintf("{\n\t\"acknowledged\": %v,\n\t\"insertedId\": \"%v\",\n\t\"matchedCount\": %d,\n\t\"modifiedCount\": %d,\n\t\"upsertedCount\": 0\n}",
			r.MatchedCount > 0, r.UpsertedID, r.MatchedCount, r.ModifiedCount)
	case *mongo.DeleteResult:
		content = fmt.Sprintf("AcknowledgedDeleteResult{deletedCount=%d}", r.DeletedCount)
	default:
		content = "Unsupported result type"
	}

	writeToOutputFile(queryStr, content, filePath)
}
